src/modules/sync/advancedDiff.ts

```diff
@@ -0,0 +1,529 @@
+  isObjectReferenceArray,
+  isTaxonomyRequestModel,
+  isValidationRegex,
+} from "../../utils/typeguards.js";
+import { RequiredCodename } from "../../utils/types.js";
+import { DiffModel, DiffObject } from "./types/diffModel.js";
+import { PatchOperation, ReplacePatchOperationValue } from "./types/patchOperation.js";
+
+type DiffData = DiffModel & SyncParams;
+type EntityPathHandler = {
+  regex: RegExp;
+  entity: (() => string) | ((match: string[]) => string);
+};
+
+export const writeDiffToFile = (diffData: DiffData) => {
+  const template = readFileSync(resolve("./", "diffTemplate.html"), "utf-8");
+  const resolvedTemplate = template.replace(
+    /{{.*?}}/g,
+    (match) => templateHandlerMap.get(match)?.(diffData) ?? match,
+  );
+  const outputPath = resolve(
+    diffData.outPath!,
+    `diff_${new Date().toISOString().replace(/[:.-]/g, "_")}.html`,
+  );
+
+  try {
+    logInfo(
+      diffData,
+      "standard",
+      chalk.yellow(`Generating a diff file at ${outputPath}`),
+    );
+    writeFileSync(outputPath, resolvedTemplate);
+  } catch (err) {
+    logError(diffData, `Failed writing a diff file: ${JSON.stringify(err)}`);
+    process.exit(1);
+  }
+
+  logInfo(
+    diffData,
+    "standard",
+    chalk.green(`Diff file created successfully. Opening...`),
+  );
+  open(outputPath).catch((err) =>
+    logError(
+      diffData,
+      `Failed to open the file: ${outputPath}\nMessage: ${JSON.stringify(err)}`,
+    )
+  );
+};
+
+const processPatchOp = (patchOp: PatchOperation) => {
+  switch (patchOp.op) {
+    case "remove":
+      return `${getRemoveEntityDetail(patchOp.path)} removed`;
+    case "move": {
+      const beforeText = "before" in patchOp
+        ? `before <strong>${patchOp.before.codename}</strong>`
+        : "";
+      const afterText = "after" in patchOp
+        ? `after <strong>${patchOp.after.codename}</strong>`
+        : "";
+      return `${
+        getMoveEntityDetail(
+          patchOp.path,
+        )
+      } moved ${beforeText}${afterText}`;
+    }
+    case "addInto":
+      return `${getAddEntityDetail(patchOp.path)} ${
+        getValueOrIdentifier(
+          patchOp.value,
+        )
+      } added ${patchOp.before ? `before ${patchOp.before.codename}` : ""}${
+        patchOp.after ? `after ${patchOp.after.codename}` : ""
+      }`;
+    case "replace":
+      return `${
+        getReplaceEntityDetail(
+          patchOp.path,
+        )
+      } changed <div class="compared-elements"><div class="compared-element">${
+        getReplaceOpValue(
+          patchOp.oldValue,
+        )
+      }</div><div class="comparator">→</div><div class="compared-element"> ${
+        getReplaceOpValue(
+          patchOp.value,
+        )
+      }</div></div>`;
+  }
+};
+
+const getEntityPathHandler = (handlers: EntityPathHandler[], path: string) => {
+  const pattern = handlers.find(h => h.regex.test(path));
+
+  return pattern
+    ? pattern.entity(path.match(pattern.regex) as string[])
```

Comment:
true, good catch. that being said, I'm not sure the way I refactored it is completely correct either. have a look if it's alright with you. 

I don't think it's typed correctly since some of the render methods have no arguments (not an empty array)